app/services/invoices/aggregate_amounts_and_taxes_from_fees.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+module Invoices
+  class AggregateAmountsAndTaxesFromFees < BaseService
+    Result = BaseResult
+
+    def initialize(invoice:)
+      @invoice = invoice
+
+      raise ArgumentError.new("invoice type must be `advance_charges`") unless invoice.advance_charges?
```

Comment:
Errors returned in the results are error that can happen. Here I want to ensure that if another dev wants to use this service with another type of invoice, they are warned the service wasn't designed for this and they should make changes.

Reply:
I think it is still a precondition, I would stick to returning a failed result.
It should then be raised/handled by the caller.  We already have many similar situations

The risk with using "ArgumentError" or other built-in ruby error is that for some reason, someone could catch the error type in higher services with some custom handling logic.

Reply:
Yes, I wasn't too sure about this myself